pyro/params/param_store.py

```diff
@@ -46,18 +49,133 @@ def clear(self):
         self._param_to_name = {}
         self._constraints = {}
 
+    def items(self):
+        """
+        Iterate over ``(name, constrained_param)`` pairs.
+        """
+        for name in self._params:
+            yield name, self[name]
+
+    def keys(self):
+        """
+        Iterate over param names.
+        """
+        return self._params.keys()
+
+    def values(self):
+        """
+        Iterate over constrained parameter values.
+        """
+        for name, constrained_param in self.items():
+            yield constrained_param
+
+    def __bool__(self):
+        return bool(self._params)
+
+    def __len__(self):
+        return len(self._params)
+
+    def __contains__(self, name):
+        return name in self._params
+
+    def __iter__(self):
+        """
+        Iterate over param names.
+        """
+        return iter(self.keys())
+
+    def __delitem__(self, name):
+        """
+        Remove a parameter from the param store.
+        """
+        constrained_value = self._params.pop(name)
+        unconstrained_value = constrained_value.unconstrained()
+        self._param_to_name.pop(unconstrained_value)
+        self._constraints.pop(name)
+
+    def __getitem__(self, name):
+        """
+        Get the constrained value of a named parameter.
+        """
+        unconstrained_value = self._params[name]
+
+        # compute the constrained value
+        constraint = self._constraints[name]
+        constrained_value = transform_to(constraint)(unconstrained_value)
+        constrained_value.unconstrained = weakref.ref(unconstrained_value)
+
+        return constrained_value
+
+    def __setitem__(self, name, new_constrained_value):
+        """
+        Set the constrained value of an existing parameter, or the value of a
+        new unconstrained parameter. To declare a new parameter with
+        constraint, use :meth:`setdefault`.
+        """
+        # store constraint, defaulting to unconstrained
+        constraint = self._constraints.setdefault(name, constraints.real)
+
+        # compute the unconstrained value
+        with torch.no_grad():
+            # FIXME should we .detach() the new_constrained_value?
```

Comment:
I tried to `.detach()` here but gave up due to failures in `test_em_nested_in_svi` and `tests/integration_tests/test_tracegraph_elbo.py::RaoBlackwellizationTests.test_iarange_in_elbo_with_superfluous_rvs`. I wasn't able to determine the reason behind the latter failure, so I'm avoiding the unnecessary change.

Reply:
bugs 1 
fritz 0